Add support for resizable overlays

deathbeam opened the pull request:
Add mostly complete support for resizable overlays.

Closes #3916
Original PR: #4509

Signed-off-by: Tomas Slusny <slusnucky@gmail.com>

Preview:
![peek 2018-07-26 21-02](https://user-images.githubusercontent.com/5115805/43282844-843325a4-9117-11e8-865b-42ae5c6f1539.gif)

Adam- commented on runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayPanel.java:
```diff
@@ -0,0 +1,105 @@
+	 * Overlay name, used for saving the overlay, needs to be unique
+	 * @return overlay name
+	 */
+	public String getName()
```
I think this can just be deleted

abextm commented on runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java:
```diff
@@ -277,9 +287,21 @@ public void render(Graphics2D graphics, final OverlayLayer layer)
-						graphics.setColor(movedOverlay == overlay ? MOVING_OVERLAY_ACTIVE_COLOR : MOVING_OVERLAY_COLOR);
+						final Stroke previousStroke = graphics.getStroke();
+
+						if (hoveredOverlay == overlay && !clientUI.getCurrentCursor().equals(clientUI.getDefaultCursor()))
```
this should probably be a reference comparison, or better yet, a flag for the resizing mode rather than reusing the cursor

abextm commented on runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayPanel.java:
```diff
@@ -0,0 +1,96 @@
+	private OverlayPosition position = OverlayPosition.TOP_LEFT;
+	private OverlayPriority priority = OverlayPriority.NONE;
+	private OverlayLayer layer = OverlayLayer.UNDER_WIDGETS;
+	private boolean clearChildren = true;
```
it may be worth adding some javadoc to these

abextm commented on runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayPanel.java:
```diff
@@ -0,0 +1,96 @@
+public abstract class OverlayPanel extends Overlay
+{
+	protected final PanelComponent panelComponent = new PanelComponent();
+	private Point preferredLocation;
```
why are you shadowing a bunch of stuff from Overlay?

deathbeam commented:
Added javadoc, removed that shadowed stuff & added docs

abextm commented:
If you resize an overlay in a hotspot it will be moved out of the hotspot, but still be in the same spot, under other overlays.

deathbeam commented:
Fixed. Also fixed small issue with stuff jumping around while resizing (its not jumping around at all anymore)

abextm commented:
If you try to resize a WidgetOverlay it will offset it

deathbeam commented:
Hmmmmm. Maybe I should just add some isResizable flag to Overlay, set it to true in OverlayPanel and also in ScreenMarker overlay and just ignore rest. I think that would be safest approach?

deathbeam commented:
Alright, updated with that

abextm commented on runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java:
```diff
@@ -294,125 +314,262 @@ public void render(Graphics2D graphics, final OverlayLayer layer)
+			case Rectangle.OUT_TOP:
+				clientUI.setCursor(Cursor.getPredefinedCursor(Cursor.N_RESIZE_CURSOR));
+				break;
+			case Rectangle.OUT_TOP + Rectangle.OUT_LEFT:
```
use bitwise or for oring bitfields

abextm commented on runelite-client/src/main/java/net/runelite/client/ui/overlay/OverlayRenderer.java:
```diff
@@ -68,21 +70,29 @@
 	private static final int BORDER_TOP = BORDER + 15;
 	private static final int PADDING = 2;
+	private static final int MIN_OVERLAY_SIZE = 32;
+	private static final int OVERLAY_RESIZE_TOLERANCE = 10;
```
It might be good to drop this a few px, its easy to get the resize zone on small overlays when you want to move it.

deathbeam commented:
Updated, changed the tolerance to 5px

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java:
```diff
@@ -57,8 +57,9 @@
-		panelComponent.setWrapping(4);
+		panelComponent.setWrap(true);
 		panelComponent.setGap(new Point(6, 4));
+		panelComponent.setPreferredSize(new Dimension(4 * Constants.ITEM_SPRITE_WIDTH, 0));
```
This also include the padding/gap, does that matter?

Adam- commented on runelite-client/src/main/java/net/runelite/client/ui/overlay/infobox/InfoBoxOverlay.java:
```diff
@@ -83,11 +84,10 @@ public Dimension render(Graphics2D graphics)
 
 		panelComponent.getChildren().clear();
-		panelComponent.setWrapping(config.infoBoxWrap());
+		panelComponent.setPreferredSize(new Dimension(config.infoBoxSize(), config.infoBoxSize()));
```
This appears to be causing the infoboxes to wrap at 2 (although it reads like it would make it wrap after 1..?). Regardless it should stay consistent with the current behavior and wrap at 4 by default.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/npcunaggroarea/NpcAggroAreaNotWorkingOverlay.java:
```diff
@@ -27,30 +27,22 @@
 
-		panelComponent = new PanelComponent();
-		panelComponent.setPreferredSize(new Dimension(150, 0));
-		panelComponent.getChildren().add(LineComponent.builder()
```
Couldn't you have instead just `setClearChildren(false)`?

deathbeam commented:
Updated. Good catch with the gap. Now infoboxes are wrapped at 4 by default, I updated the description of input consuming commit and reverted the change in NpcAggroOverlay and changed it to use setClearChildren(false) instead.

Adam- commented:
This still does not wrap at 4 infoboxes for me, but now 5, by default.
